### ISSUE TITLE
TypeError: Got unexpected return type 'Future' in RedisProtocol.decr, expected <class 'int'>

### ISSUE THREAD

**Issue by adamrothman** [jonathanslenders/asyncio-redis#98]
This has been happening with increasing frequency in production for us:

```
TypeError: Got unexpected return type 'Future' in RedisProtocol.decr, expected <class 'int'>
  File "courier/handlers/socket.py", line 271, in handle_close
    connection_count = await self._decrement_open_connections()
  File "courier/handlers/socket.py", line 169, in _decrement_open_connections
    decr_result = await self.redis.decr(self._open_connections_key)
  File "asyncio_redis/protocol.py", line 659, in wrapper
    typecheck_return(protocol_self, result)
  File "asyncio_redis/protocol.py", line 517, in typecheck_return
    (type(result).__name__, self.method.__name__, expected_type))
```

It doesn't happen _every_ time, but it definitely happens enough to be a problem. It seems to be a bug in this library. @jonathanslenders any idea why the protocol is sometimes returning a `Future` and sometimes an `int`?

Our stack is:

- Python 3.5.2
- `tornado` 4.4.1 (using the `asyncio` event loop)
- `asyncio_redis` 0.14.2
- Redis 3.2.3

**Comment by jonathanslenders** [comment]
Hi @adamrothman,

This is clearly a bug. Thanks for reporting it!
Right now, I have no idea yet what's going on, but I'm looking into it.

**Comment by jonathanslenders** [comment]
One question. Are you using transactions?
If so, the bug could be related to that. (If not, I know I have to look into something else. -- It's not that obvious.)

**Comment by adamrothman** [comment]
Not in this case, no (but I have another stack trace I can open an issue for that I think _is_ related to transactions). This is the source for the `_decrement_open_connections`:

```
async def _decrement_open_connections(self):
    decr_result = await self.redis.decr(self._open_connections_key)
    if decr_result < 0:
        # Key doesn't exist (or it expired); delete it
        await self.redis.delete([self._open_connections_key])
        return 0
    return decr_result
```

**Comment by adamrothman** [comment]
@jonathanslenders done (#99); thanks for looking into these!

**Comment by jonathanslenders** [comment]
Hi @adamrothman,

I could not reproduce any of the above issues, but could reproduce a very similar condition. Apparently, transactions were prone to race conditions, when they were started from different coroutines. (Which happened once in a while.)

This fix was not as easy as I thought in the first place, but I think, I have a working fix:
https://github.com/jonathanslenders/asyncio-redis/pull/100
Basically, I had to use an ``asyncio.Lock`` at the right places.

In my tests, no crashes happened anymore, and all the unit tests succeed. Do you have any possibility to test this branch?

**Comment by adamrothman** [comment]
@jonathanslenders it makes sense to me that your fix would address #99, but are you saying it solves this issue as well? We should be able to test out your branch; I'll deploy it tomorrow and keep you posted on the results.

**Comment by jonathanslenders** [comment]
Yes, I think that it does and they are caused by the same issue. Not 100% sure, but it definitely makes sense to me. Let me know if it works or not.

**Comment by adamrothman** [comment]
@jonathanslenders We have been running your patched branch since Monday – haven't seen any of this crash or the other. I'll leave it up to you whether to close the issues or seek further confirmation.

**Comment by adamrothman** [comment]
@jonathanslenders I am satisfied that the patch has fixed our issues. I'll close this and the other issue – feel free to merge #100. Any idea when the patch will make it into an official release?

**Comment by adamrothman** [comment]
Er, actually, I'm going to leave these open until the branch is merged.

**Comment by jonathanslenders** [comment]
Hi @adamrothman,
I will try to publish a new release this weekend.
Thanks again for reporting the issue and trying out the fix!

**Comment by jonathanslenders** [comment]
Hi @adamrothman,

#100 has been merged and it has been released on Pypi: https://pypi.python.org/pypi/asyncio_redis/

So closing this. Please let me know if you experience further issues.
